Non determinism in test_jsongenerator

```
coverage run -p --source=dss tests/test_jsongenerator.py -v
ok
test_file_put_large (__main__.TestFileApi) ... skipped 'Skipping integration test'
test_file_put_metadata_from_tags (__main__.TestFileApi) ... test_additionalItems (__main__.TestArray) ... ok
test_common (__main__.TestArray) ... ok
test_file_put_upper_case_checksums (__main__.TestFileApi) ... ok
test_constants (__main__.TestArray) ... ok
test_contains (__main__.TestArray) ... ok
test_enums (__main__.TestArray) ... ok
test_in_range (__main__.TestArray) ... ok
test_items_is_array (__main__.TestArray) ... ok
test_items_is_schema (__main__.TestArray) ... ok
test_unique (__main__.TestArray) ... ok
test_unique_enums (__main__.TestArray) ... ok
test_common (__main__.TestInteger) ... ok
test_exclusive_range (__main__.TestInteger)
does not exceed exclusive range ... ok
test_inclusive_range (__main__.TestInteger) ... ok
test_multiple_of (__main__.TestInteger) ... ok
test_multiple_of_in_range (__main__.TestInteger)
is a multiple of and within boundaries ... ok
test_generate (__main__.TestJsonGenerator) ... ok
test_file_size (__main__.TestFileApi) ... ok
test_common (__main__.TestNumber) ... ok
test_exclusive_range (__main__.TestNumber) ... ok
test_inclusive_range (__main__.TestNumber) ... test_multiple_of (__main__.TestNumber) ... ok
test_multiple_of_in_inclusive_range (__main__.TestNumber) ... ok
test_additionalProperties (__main__.TestObject) ... ok
test_dependencies (__main__.TestObject) ... ok
test_patternProperties (__main__.TestObject) ... ok
test_properties (__main__.TestObject) ... ok
test_propertyNames (__main__.TestObject) ... ok
test_range (__main__.TestObject) ... ok
test_required (__main__.TestObject) ... ok
test_common (__main__.TestString) ... ok
test_maxLength (__main__.TestString)
does not exceed maxLength ... ok
test_minLength (__main__.TestString)
does not exceed minLength ... ok
test_minimum_is_maximum (__main__.TestString)
minLength == maxLength ... ok
test_pattern (__main__.TestString) ... ok
test_pattern_within_range (__main__.TestString)
WIP matches pattern and does not exceed minLength or maxLength ... ok
test_range (__main__.TestString)
does not exceed maxLength or minLength ... ok
======================================================================
FAIL: test_inclusive_range (__main__.TestNumber) [does not exceed minimum of 0.1]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_jsongenerator.py", line 257, in test_inclusive_range
    self.assertGreaterEqual(self.json_gen._number(schema), minimum)
AssertionError: 0.0 not greater than or equal to 0.1
----------------------------------------------------------------------
Ran 35 tests in 4.002s
FAILED (failures=1)
Makefile:39: recipe for target 'tests/test_jsongenerator.py' failed
make: *** [tests/test_jsongenerator.py] Error 1
```

--- a/dss/util/json_gen/generator.py
+++ b/dss/util/json_gen/generator.py
@@ -5,6 +5,7 @@ import rstr
 from faker import Faker
 from faker.providers.python import Provider as PythonProvider
 from jsonschema import RefResolver, Draft4Validator
+from math import ceil, floor
 
 
 class JsonProvider(PythonProvider):
@@ -184,17 +185,19 @@ class JsonGenerator(object):
     def _number(self, schema: dict) -> Union[int, float]:
         impostor = self._common(schema)
         if impostor is None:
-            maximum = schema.get('maximum', schema.get('exclusiveMaximum', self.UNBOUND_MAX_INT) - 1)
-            minimum = schema.get('minimum', schema.get('exclusiveMinimum', self.UNBOUND_MIN_INT) + 1)
+            maximum = schema.get('maximum', schema.get('exclusiveMaximum', self.UNBOUND_MAX_INT) - 1e-12)
+            minimum = schema.get('minimum', schema.get('exclusiveMinimum', self.UNBOUND_MIN_INT) + 1e-12)
             if minimum == maximum:
                 impostor = minimum
             else:
-                multiple_of = schema.get('multipleOf', 1)
-                if isinstance(maximum, float) or isinstance(minimum, float) or isinstance(minimum, float):
-                    impostor = self.faker.random.uniform(minimum, maximum)
+                multiple_of = schema.get('multipleOf')
+                if multiple_of is not None:
+                    if multiple_of <= 0:
+                        raise ValueError("multipleOf must be > 0")
+                    v = self.faker.random_int(ceil(minimum / multiple_of), floor(maximum / multiple_of))
+                    impostor = round(v * multiple_of, 12)
                 else:
-                    impostor = self.faker.random_int(minimum, maximum)
-                impostor = impostor // multiple_of * multiple_of
+                    impostor = self.faker.random.uniform(minimum, maximum)
         return impostor
 
     def _integer(self, schema: dict) -> int:
@@ -205,9 +208,14 @@ class JsonGenerator(object):
             if minimum == maximum:
                 impostor = minimum
             else:
-                impostor = self.faker.random_int(minimum, maximum)
-                multiple_of = schema.get('multipleOf', 1)
-                impostor = max(minimum, impostor // multiple_of * multiple_of)
+                multiple_of = schema.get('multipleOf')
+                if multiple_of is not None:
+                    if multiple_of <= 0:
+                        raise ValueError("multipleOf must be > 0")
+                    v = self.faker.random_int(ceil(minimum / multiple_of), floor(maximum / multiple_of))
+                    impostor = v * multiple_of
+                else:
+                    impostor = self.faker.random_int(minimum, maximum)
         return impostor
 
     def _string(self, schema: dict) -> str:

--- a/tests/test_jsongenerator.py
+++ b/tests/test_jsongenerator.py
@@ -224,7 +224,7 @@ simple_object = {'type': 'object', 'properties': {'thing_1': simple_integer, 'th
 
 
 class Base(unittest.TestCase):
-    repeat = 25
+    repeat = 1000
     def setUp(self):
         self.json_gen = JsonGenerator()
 
@@ -245,8 +245,9 @@ class Base(unittest.TestCase):
 
 @testmode.standalone
 class TestNumber(Base):
-    numbers = [10.0, 0.1, 0.0, -0.1, -10.0]
-    multiple_ofs = [1.5, 2]  # TODO support multipleOf [0.1, -0.2] for floats
+    numbers = [10.0, 0.3, 0.0, -0.1, -10.0, 1e-09]
+    multiple_ofs = [0.1, 1.01, 1.5, 1.99, 2, 3]
+    invalid_multiple_ofs = [0, -2]
 
     def test_common(self):
         self._test_common(self.json_gen._number, 'number', self.numbers, 0.999)
@@ -286,19 +287,28 @@ class TestNumber(Base):
             with self.subTest(f"is a multiple of {multiple_of}"):
                 schema = {'type': 'number', 'multipleOf': multiple_of}
                 for i in range(self.repeat):
-                    self.assertFalse(self.json_gen._number(schema) % multiple_of)
+                    value = self.json_gen._number(schema)
+                    self.assertTrue(round(value / multiple_of, 3).is_integer(), msg=f"value:{value} is not a "
+                                                                                    f"multipleOf:{multiple_of}.")
+        for multiple_of in self.invalid_multiple_ofs:
+            with self.subTest(f"is a multiple of {multiple_of}"):
+                schema = {'type': 'number', 'multipleOf': multiple_of}
+                with self.assertRaises(ValueError):
+                    self.json_gen._number(schema)
 
     def test_multiple_of_in_inclusive_range(self):
-        maximum = 5.0
+        maximum = 10.0
         minimum = 0.0
         for multiple_of in self.multiple_ofs:
             with self.subTest(f"is a multiple of {multiple_of} and within boundaries"):
                 schema = {'type': 'number', 'minimum': minimum, 'maximum': maximum, 'multipleOf': multiple_of}
                 for i in range(self.repeat):
                     value = self.json_gen._number(schema)
+                    v = round(value / multiple_of, 3)
                     self.assertLessEqual(value, maximum)
                     self.assertGreaterEqual(value, minimum)
-                    self.assertFalse(value % multiple_of)
+                    self.assertTrue(v.is_integer(), msg=f"value:{value} is not a "
+                                                        f"multipleOf:{multiple_of}.")
 
     def test_exclusive_range(self):
         with self.subTest("does not exceed exclusive range"):
@@ -307,7 +317,8 @@ class TestNumber(Base):
             schema = {'type': 'number', 'exclusiveMinimum': minimum, 'exclusiveMaximum': maximum}
             for i in range(self.repeat):
                 value = self.json_gen._number(schema)
-                self.assertEqual(value, 2)
+                self.assertLess(value, maximum)
+                self.assertGreater(value, minimum)
 
         with self.subTest("WIP minimum == maximum"):
             if False:
@@ -354,12 +365,19 @@ class TestInteger(Base):
                 self.assertEqual(value, 1)
 
     def test_multiple_of(self):
-        multiple_ofs = [5, -5]
+        multiple_ofs = [2, 3, 5, 7]
         for multiple_of in multiple_ofs:
             with self.subTest(f"is a multiple of {multiple_of}"):
                 schema = {'type': 'integer', 'multipleOf': multiple_of}
                 for i in range(self.repeat):
                     self.assertFalse(self.json_gen._integer(schema) % multiple_of)
+
+        invalid_multiple_ofs = [0, -2]
+        for multiple_of in invalid_multiple_ofs:
+            with self.subTest(f"is a multiple of {multiple_of}"):
+                schema = {'type': 'number', 'multipleOf': multiple_of}
+                with self.assertRaises(ValueError):
+                    self.json_gen._integer(schema)
 
     def test_multiple_of_in_range(self):
         """is a multiple of and within boundaries"""

